aten/src/ATen/native/sparse/SparseUtils.h

```diff
@@ -0,0 +1,127 @@
+#include <ATen/ATen.h>
+#include <ATen/SparseTensorImpl.h>
+
+namespace at { namespace native {
+
+// Just for documentary purposes
+using SparseTensor = Tensor;
+using LongTensor = Tensor;
+using IntTensor = Tensor;
+using SparseType = Type;
+
+namespace {
+
+// This is an internal utility function for getting at the SparseTensorImpl,
+// so that we can write sparse tensor specific accessors for special fields
+// in SparseTensor.  You should only use this for writing low level
+// setters/getters for SparseTensorImpl fields; otherwise, you should use
+// the low level setters/getters that were implemented using this.
+//
+// This may be called repeatedly, so make sure it's pretty cheap.
+SparseTensorImpl* _get_sparse_impl(const SparseTensor& self) {
+  if (!self.is_sparse()) AT_ERROR("_internal_get_SparseTensorImpl: not a sparse tensor");
+  return static_cast<SparseTensorImpl*>(self.unsafeGetTensorImpl());
+}
+
+// Port of the old THCSTensor_(checkGPU), but it doesn't really belong here
+// because it is more general
+// NB: I dropped kernelP2PEnabled support
+// NB: This does NOT assume that the tensors are GPU tensors; if they're
+// all CPU that's OK too
+inline bool _check_device(ArrayRef<Tensor> ts) {
+  if (ts.empty()) {
+    return true;
+  }
+  const Tensor& ref_t = ts.front();
+  int64_t curDevice = -1;
+  if (ref_t.is_cuda()) {
+    curDevice = current_device();
+  }
+  for (const Tensor& t : ts) {
+    if (t.get_device() != curDevice) return false;
```

Comment:
wait, `get_device` works on CPU devices?  How?

Reply:
It doesn't, the comment is wrong.